extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/interceptor/TransactionalInterceptorBase.java

```diff
@@ -86,8 +96,31 @@ protected Object invokeInOurTx(InvocationContext ic, TransactionManager tm) thro
     protected Object invokeInOurTx(InvocationContext ic, TransactionManager tm, RunnableWithException afterEndTransaction)
             throws Exception {
 
-        tm.begin();
-        Transaction tx = tm.getTransaction();
+        TransactionConfiguration configAnnotation = getTransactionConfiguration(ic);
+        int currentTmTimeout = ((TransactionManagerImple) com.arjuna.ats.jta.TransactionManager.transactionManager())
+                .getTimeout();
+        if (configAnnotation != null && configAnnotation.timeout() != TransactionConfiguration.UNSET_TIMEOUT) {
+            tm.setTransactionTimeout(configAnnotation.timeout());
+        }
+        Transaction tx;
+        try {
+            tm.begin();
+            tx = tm.getTransaction();
+            if (tx.getStatus() == Status.STATUS_NO_TRANSACTION && configAnnotation != null) {
```

Comment:
I might have misunderstood but this seems impossible? I.e. that tm.begin() would succeed but subsequently STATUS_NO_TRANSACTION would be found?

Reply:
Right, I think I confused @loicmathieu with my explanations. I gave a clearer explanation (I hope  ;) ) here
https://github.com/quarkusio/quarkus/commit/e093f3735d1c8fd45262524723a213ee67140ee3

In particular

@loicmathieu this is not the correct place.
If you look at TransactionalInterceptorBase and its subclasses e.g. TransactionalInterceptorRequired you will see that the method called during interception when a transaction has already been started (and therefore when we are called from a method tha already had @Transactional or equivalent) is not invokeInOurTx but invokeInCallerTx (and invokeInNoTx in the case of TransactionalInterceptorNever, TransactionalInterceptorNotSupported or TransactionalInterceptotSupports.

So the code should look something like that.

```
private void checkConfiguration(InvocationContext ic) {
    TransactionConfiguration configAnnotation = getTransactionConfiguration(ic);
    if (configAnnotation != null && configAnnotation.timeout() != TransactionConfiguration.UNSET_TIMEOUT) {
        throw new TransactionalException("Cannot change timeout via @TransactionConfiguration in an already started transaction: you must ensure that the annotation is set on the method that creates the transaction.");
    }
    return;
}

protected Object invokeInNoTx(InvocationContext ic) throws Exception {
    checkConfiguration(ic);
    return ic.proceed();
}


protected Object invokeInCallerTx(InvocationContext ic, Transaction tx) throws Exception {
    try {
        checkConfiguration(ic);
        return ic.proceed();
    } catch (Exception e) {
        handleException(ic, e, tx);
    }
    throw new RuntimeException("UNREACHABLE");
}
```

Am I making sense? We can arrange a call otherwise as it might be easier to explain by voice. (I'll not be available in the next two days though).

Reply:
@emmanuelbernard yes, you confused me yesterday and as I make the change too quickly I didn't saw that you talked about `invokeInCallerTx`.

I re-read the all story and the code, and now I understand, and it makes more sense as I didn't see how it can works in the first place (so I must admit that I update the code without really understanding it ... playing real-life day-to-day developement here ;) ).

So, I implements what you suggested.